src/uct/ib/mlx5/ib_mlx5.inl

```diff
@@ -72,13 +72,119 @@ uct_ib_mlx5_update_db_cq_ci(uct_ib_mlx5_cq_t *cq)
+        host_sop_drop_qpn  &= UCS_MASK(UCT_IB_QPN_ORDER);
+        host_sop_drop_qpn  |= host_s_wqe_opcode << UCT_IB_QPN_ORDER;
+        dst->sop_drop_qpn   = htonl(host_sop_drop_qpn);
+    }
+}
+
+
+/**
+ * Continue decompressing the CQEs. Should be used only when cq_decomp->left > 0.
+ * 
+ * @return        Next decompressed CQE.
+ *
+ * @param cq      CQ that contains unhandled compressed CQEs (cq_decomp->left > 0).
+ * @param is_rx   Defines the CQ side (RX or TX).
+ *
+ */
+static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
+uct_ib_mlx5_iface_cqe_decomp_continue(uct_ib_mlx5_cq_t *cq, int is_rx) {
+    uct_ib_mlx5_cq_decomp_t *cq_decomp = &cq->cq_decomp;
+    struct mlx5_cqe64 *title           = &cq_decomp->title;
+    uct_ib_mlx5_mini_cqe8_t *mini_cqe  = &cq_decomp->mini_arr[cq_decomp->mini_arr_idx % 8];
+    struct mlx5_cqe64 *cqe;
+
+    /* Update opcode in the corresponging CQE buffer*/
+    cqe = uct_ib_mlx5_get_cqe(cq, cq->cq_ci);
+    cqe->op_own = title->op_own;
+
+    uct_ib_mlx5_iface_cqe_decomp_fill_unique_fields(title, mini_cqe,
+                                                    cq_decomp, is_rx);
+
+    cq_decomp->left--;
+    cq_decomp->mini_arr_idx++;
+    cq->cq_ci += 1;
+
+    if ((cq_decomp->left != 0) && (cq_decomp->mini_arr_idx % 8 == 0)) {
+        /* Get the next mini_cqe array */
+        cqe = uct_ib_mlx5_get_cqe(cq, cq->cq_ci);
+        memcpy(&cq_decomp->mini_arr, cqe, sizeof(*cqe));
+    }
+
+    return title;
+}
+
+
+/**
+ * Initialize the CQE decompression related entities and decompress the first CQE.
+ * 
+ * @return        First decompressed cqe.
+ *
+ * @param title   The CQE that contains the title of the compression block.
+ * @param cq      CQ that contains the title.
+ * @param is_rx   Defines the CQ side (RX or TX).
+ *
+ */
+static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
+uct_ib_mlx5_iface_cqe_decomp_start(struct mlx5_cqe64 *title,
+                                   uct_ib_mlx5_cq_t *cq, int is_rx) {
+    uct_ib_mlx5_cq_decomp_t *cq_decomp = &cq->cq_decomp;
+    struct mlx5_cqe64 *mini_cqe        = uct_ib_mlx5_get_cqe(cq, cq->cq_ci + 1);
+
+    cq_decomp->block_size   = ntohl(title->byte_cnt);
+    cq_decomp->left         = cq_decomp->block_size;
+    cq_decomp->wqe_counter  = ntohs(title->wqe_counter);
+    cq_decomp->mini_arr_idx = 0;
+    memcpy(&cq_decomp->title,    title, sizeof(struct mlx5_cqe64));
+    memcpy(&cq_decomp->mini_arr, mini_cqe,  sizeof(struct mlx5_cqe64));
+
+    UCT_IB_MLX5_CQE_FORMAT_CLEAR(cq_decomp->title.op_own);
+    return uct_ib_mlx5_iface_cqe_decomp_continue(cq, is_rx);
+}
+
+
 static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
-uct_ib_mlx5_poll_cq(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq)
+uct_ib_mlx5_poll_cq(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq, int is_rx)
 {
     struct mlx5_cqe64 *cqe;
     unsigned cqe_index;
     uint8_t op_own;
 
+    if (cq->cq_decomp.left > 0) {
+        return uct_ib_mlx5_iface_cqe_decomp_continue(cq, is_rx);
```

Comment:
I don't think we should decompress CQEs and then read them from CQE buffer, like the mlx5 driver is doing. A more efficient implementation could be to parse the mini cqe and process it as completion, without decompressing it to an array of cqe64. 
If this is too complex for initial implementation, at least we need to decompress all mini_cqes as soon as we see them. this way could remove this branch, and no need to copy min_cqe_arr.

Reply:
Discussed orally, please reopen in case of questions.